.github/workflows/docker.yml

```diff
@@ -0,0 +1,80 @@
+name: (sub) Build Docker image
+
+defaults:
+  run:
+    shell: bash
+
+on:
+  workflow_call:
+    inputs:
+      publish:
+        description: A flag that triggers publishing to GHCR
+        type: boolean
+        default: false
+      python_version:
+        description: A version of Python to install
+        type: string
+        required: true
+
+jobs:
+
+  build:
+    name: Build ${{ inputs.publish && 'and publish' || '' }} a Docker image
+    runs-on: ubuntu-latest
+    permissions:
+      packages: write
+      contents: read
+    steps:
+      - name: checkout code
+        uses: actions/checkout@v3
+
+      - name: make sure we pull the latest commits
+        if: inputs.publish
+        run: git pull
```

Comment:
this is only when we run a publish job, so during a release. Trigger for that job is a schedule. So for schedules, when you run a `checkout` job it checks out a commit which was the latest commit when the job triggered. 

For releases, a `semantic-release` step calculates a new release version, we update that version in `pyproject.toml` and when a new release is pushed we commit that version directly to the `main` branch. The commit is done automatically. 

Now, if we run `checkout` after a release (which we do) we will miss that commit. And we need it, as we run `poetry version -s` command in line 46 to get the actual release version to tag the image correctly. Hence we run `git pull` to get the latest commit. 

> This might also bring the possibility that some other commits get pushed/merged after release workflow is run

Highly unlikely to happen:
- as a release run at 1am
- the `pull` command is run in a chain 
- it's only us that have access to commit to the `main` branch.

Reply:
@alperenkose - `git pull` replaced with a particular ref checkout, tested, works just fine. 

For PRs, when you leave the `ref` property empty it checks out the head of the PR branch by default